### PR TITLE
Re-organize project to improve readability, documentation

### PR DESCRIPTION
**Pull request by gselzer**
This PR does a couple major things:
* File renaming and splitting: Some files were poorly named (like `_ptypes.py`). They now have better names, and are often in subdirectories (e.g. `_ptypes.py` -> `types/mappings.py`. Other files were split into multiple files; this was the case for `widget.py`, which contained the main napari-imagej widget and all subwidgets. `widget.py` was moved into a new `widgets` directory and was split into multiple files.
* File documentation: All files should now have a module docstring specifying its purpose and highlighted usages.

Tests were also moved, but should not be altered.
* Exception* There was one test, testing that the napari-imagej widget was vertically oriented. Since this is actually tested in another test in that file, that test was deleted.

This PR doesn't actually close any issues, but it polishes the code a great deal in anticipation of the 0.1.0 release.

### CODE REVIEW

File: tests/types/test_type_conversions.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+from typing import List
+
+import pytest
+
+from napari_imagej.types.mappings import ptypes
+from napari_imagej.types.placeholders import OutOfBoundsFactory
```
Comment (gselzer) [inline tests/types/test_type_conversions.py:6]: Might want to add a short module docstring at the top

Comment (gselzer) [reply tests/types/test_type_conversions.py:6]: And do the same for all of the test modules

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    @JavaClasses.blocking_import
+    def System(self):
+        return "java.lang.System"
+
+
+jc = JavaClassesTest()
+
+
+jc = JavaClassesTest()
+
```
Comment (gselzer) [inline tests/utils.py:54]: Duplicates 😆

File: src/napari_imagej/widgets/focuser.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
+"""
+A QWidget designed to highlight SciJava Modules.
```
Comment (gselzer) [inline src/napari_imagej/widgets/focuser.py:2]: Maybe this isn't the best wording. Choose either `focus` or `highlight`.

File: tests/widgets/test_widget.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+from napari import Viewer
```
Comment (gselzer) [inline tests/widgets/test_widget.py:1]: Maybe rename this `test_napari_imagej.py`?

Comment (gselzer) [reply tests/widgets/test_widget.py:1]: Or at least make this name the same as the file currently named `widgets/napari_imagej.py`.

File: tests/widgets/test_results_table.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
+import pytest
```
Comment (gselzer) [inline tests/widgets/test_results_table.py:1]: I'd rename this to `test_results.py`, although I'd prefer a better name for the source module...

**Review by gselzer — COMMENTED** [review]
I think this is *much* cleaner. But we can make a couple more small changes!

**Comment by gselzer** [comment]
Oh, one other thing I'd like to change, but couldn't think of a good way of doing:

I think that the file containing all of the additonal scyjava `Converter`s (currently `types/type_conversions.py`) is super cluttered, and I'd prefer to rewrite this as a set of modules, something like `types/converters/labels.py`, `types/converters/shapes.py`, etc., with a couple of decorators that would automatically register a function as a `Converter`.

...but I actually have a better idea now, so I'm going to try it :grin:

File: src/napari_imagej/types/placeholders.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    :param default: the return when we can't find a placeholder
+    :return: a placeholder for java_type, or default if we can't find one.
+    """
+    for placeholder in PLACEHOLDERS:
+        if java_type == placeholder.backing_type():
+            return placeholder
+    return default
+
+
+class StructuringElement(PythonStandin):
```
Comment (ctrueden) [inline src/napari_imagej/types/placeholders.py:58]: How about auto-generating these enum subclasses using Python's dynamic enum feature?
```python
def java_enum(java_type):
  return Enum(str(java_type.class_.getName()), {str(v): v for v in java_type.values()})

StructuringElement = java_enum(jc.StructuringElement)
```

File: src/napari_imagej/types/converters/__init__.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+"""
+A module whose submodules contain additional scyjava Converters
+that are useful within napari-imagej
+
+All submodules within this module are DYNAMICALLY IMPORTED; this allows
+automatic discovery of all Converters.
+
+The previous approach declared all Converters in one module, imported
```
Comment (ctrueden) [inline src/napari_imagej/types/converters/__init__.py:8]: I would suggest not to talk about previous approaches, which most readers are unlikely to be familiar with. Instead, just focus on how things are *laid out now*.

File: src/napari_imagej/widgets/results.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        Update children with the results stored in the SearchEventWrapper
+        """
+        self.takeChildren()
+        if results and len(results):
+            self.addChildren(results)
+        self.setText(0, f"{self.title} ({len(results)})")
+        self.setExpanded(len(results) < 10)
+
+
+class ResultsTree(QTreeWidget):
```
Comment (ctrueden) [inline src/napari_imagej/widgets/results.py:68]: The naming of `ResultsTree` for the tree, versus `ResultTreeItem` for its leaf items (setting aside the `SearcherTreeItem` non-leaf items for the moment), to be a little unfortunate. It's true that the tree is for *all results* while the leaf item is for *one result*, so the inconsistency of pluralization is logical. But I wonder if we can avoid terminology that generates this inconsistency altogether.

**Review by ctrueden — COMMENTED** [review]
Just some random thoughts, a lot about naming especially.

One technical suggestion buried in here is that we could wrap the Java enums (and enum-like things) to Python enums.

File: src/napari_imagej/types/placeholders.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+The placeholders are designed for a rather narrow use case, described as follows:
+
+Some Java functionality requires an Object implementing a particular interface.
+That Object is usually chosen from a set of established implementations,
+each created without variation (giving rise to the above constraints).
+
+In such cases, we can abstract the interface to a Placeholder and the implementations
+to enumerations of that Placeholder, as done below.
+"""
+from enum import Enum, auto
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: Fixed with 4d00b4f

File: src/napari_imagej/types/placeholders.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    """An Enum that is mocking a Java class"""
+
+    def __init__(self, _):
+        # Make ourselves aware of this enum
+        enum_type = type(self)
+        if enum_type not in PLACEHOLDERS:
+            PLACEHOLDERS.append(enum_type)
+
+    @staticmethod
+    def backing_type():
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: Fixed with ba248bb

File: src/napari_imagej/types/placeholders.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+2. An object with a no-args constructor
+
+The placeholders are designed for a rather narrow use case, described as follows:
+
+Some Java functionality requires an Object implementing a particular interface.
+That Object is usually chosen from a set of established implementations,
+each created without variation (giving rise to the above constraints).
+
+In such cases, we can abstract the interface to a Placeholder and the implementations
+to enumerations of that Placeholder, as done below.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: Added in 1f2a63f

File: src/napari_imagej/types/mappings.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+@map_category
+def images() -> Dict[Any, Any]:
+    return {
+        jc.RandomAccessibleInterval: "napari.layers.Image",
+        jc.RandomAccessible: "napari.layers.Image",
+        jc.IterableInterval: "napari.layers.Image",
+        # TODO: remove 'add_legacy=False' -> struggles with LegacyService
+        # This change is waiting on a new pyimagej release
+        # java_import('ij.ImagePlus'):
+        # 'napari.types.ImageData'
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/mappings.py:93]: Mentioned [here](https://github.com/imagej/napari-imagej/issues/94#issuecomment-1234371372)

File: src/napari_imagej/setup_imagej.py [hunk 13]
```diff
@@ -1,32 +1,88 @@
+"""
+A module designed to encapsulate all Java setup.
+
+Notable functions included in the module:
+    * ij()
+        - used to access the ImageJ instance
+    * ensure_jvm_started()
+        - used to block execution until the ImageJ instance is ready.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/setup_imagej.py]: Fixed in 6cca04b

File: src/napari_imagej/setup_imagej.py [hunk 14]
```diff
@@ -1,32 +1,88 @@
+"""
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/setup_imagej.py:1]: Renamed to `java.py` in 41745c2

File: src/napari_imagej/types/styles.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
+from typing import Dict, Optional, Union
+
+from napari_imagej.setup_imagej import jc
+
+# The definitive mapping of scyjava widget styles to magicgui widget types
+# This map allows us to determine the "best" widget for a given ModuleItem.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/styles.py]: @ctrueden what do you think of `widget_preferences.py` (ala d49d710)?

File: src/napari_imagej/types/styles.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
+from typing import Dict, Optional, Union
+
+from napari_imagej.setup_imagej import jc
+
+# The definitive mapping of scyjava widget styles to magicgui widget types
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/styles.py]: The block comment is mostly removed in d49d710

File: src/napari_imagej/types/styles.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
+"""
+The definitive mapping of scyjava widget styles to magicgui widget types
+"""
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/styles.py]: Done in d49d710, and improved in c9fac2f

File: src/napari_imagej/types/type_conversions.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+    def converter(func: Callable):
+        """Registers the annotated function with its priority"""
+        _MODULE_ITEM_CONVERTERS.append((func, priority))
+        return func
+
+    return converter
+
+
+# TODO: Move this function to scyjava.convert and/or ij.py.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/type_conversions.py]: Removed in c258c8c

File: src/napari_imagej/types/type_conversions.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+When wrapping Java functionality, we must perform data conversion in two phases.
+
+The first phase of conversion involves noting the required JAVA type of a parameter
+and determining the corresponding PYTHON type to use in harvesting input.
+
+The second phase of conversion involves transforming accepted PYTHON inputs into
+their "equivalent" JAVA objects.
+
+This module concerns itself with the first phase of conversion, while the second phase
+is left to PyImageJ's to_java function.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/type_conversions.py:13]: Documented it in c258c8c

File: src/napari_imagej/widgets/napari_imagej.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class NapariImageJ(QWidget):
+    """The top-level ImageJ widget for napari."""
+
+    def __init__(self, napari_viewer: Viewer):
+        super().__init__()
+
+        # We actually create the widgets in the opposite order,
+        # as the top widgets will want to control the ones below.
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/napari_imagej.py]: Removed in 82b9337

File: src/napari_imagej/types/styles.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ctrueden) [reply src/napari_imagej/types/styles.py]: "Preferences" implies it's something the user has control over: knobs to be turned, check boxes to be toggled. But it's not, is it? It's the canonical definition of which Java widget types go with which Python widget types. `widget_adapters` or `widget_mapping`?

[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/styles.py]: Yeah, I think `widget_mappings` is good enough. Renamed in fff6443

File: src/napari_imagej/widgets/searchbar.py [hunk 21]
```diff
@@ -0,0 +1,24 @@
+from threading import Thread
+
+from qtpy.QtWidgets import QHBoxLayout, QWidget
+
+from napari_imagej._helper_widgets import JLineEdit
+
+
+class ImageJSearchbar(QWidget):
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/searchbar.py:15]: @ctrueden I can't say I'm a huge fan of Nij :laughing:. I prefer spelling it out.

With that being said, I like the main widget being called `NapariImageJ` (although maybe we should name it `NapariImageJWidget`), and I like the menu being `NapariImageJMenu` i.e. that we specify that it is the *napari-imagej` menu.

As for the remaining widgets, I'd say that they should all either have the `NapariImageJ` prefix, or none should. I'm going to tentatively choose the latter.

So I'm going to propose some new names:
* `ImageJSearchbar` -> `JVMEnabledSearchbar`. It's basically just a searchbar that is disabled until the JVM is ready, meaning that the SciJava `Searcher`s are ready for input.
* `ResultsTree` -> `SearcherResultsTree`. This is designed to be a tree-like output to display `SearchResult`s.
* `FocusWidget` -> `SearchActionDisplay`. This is supposed to be nothing more than a label displaying the Module string and a set of buttons wrapping the different functionalities. (I dislike the word `Display`, but can't think of a better word. `SearchActionPanel`? `SearchActionWidget`?)
What do you think of these?

File: src/napari_imagej/widgets/napari_imagej.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+        self.setLayout(QVBoxLayout())
+
+        # GUI Buttons
+        self.buttons: NapariImageJMenu = NapariImageJMenu(napari_viewer)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/napari_imagej.py]: Renamed in c96f0db

File: tests/types/test_type_conversions.py (hunk 1 above)
Comment (gselzer) [reply tests/types/test_type_conversions.py]: Added in 6d71acd

File: tests/utils.py (hunk 2 above)
Comment (gselzer) [reply tests/utils.py]: Removed in f602e90

File: tests/widgets/test_widget.py (hunk 4 above)
Comment (gselzer) [reply tests/widgets/test_widget.py]: Renamed...in some commit. I don't remember which 😅

File: src/napari_imagej/widgets/searchbar.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/searchbar.py]: I performed the tentative rename in 9088810, please let me know what you think @ctrueden

File: src/napari_imagej/widgets/focuser.py [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+)
+from napari_imagej.setup_imagej import ij, jc, log_debug
+
+
+class SearchAction(NamedTuple):
+    name: str
+    action: Callable[[], None]
+
+
+class FocusWidget(QWidget):
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I think I removed them all in 50cf369

File: src/napari_imagej/widgets/focuser.py [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+
+        self.setLayout(QVBoxLayout())
+
+        self.focused_module_label = QLabel()
+        self.layout().addWidget(self.focused_module_label)
+        self.button_pane = QWidget()
+        self.button_pane.setLayout(FlowLayout())
+        self.layout().addWidget(self.button_pane)
+
+        self.focused_action_buttons = []  # type: ignore
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Did it...somewhere

File: src/napari_imagej/widgets/focuser.py [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+from napari_imagej._flow_layout import FlowLayout
+from napari_imagej._module_utils import (
+    convert_searchResult_to_info,
+    execute_function_modally,
+    functionify_module_execution,
+)
+from napari_imagej.setup_imagej import ij, jc, log_debug
+
+
+class SearchAction(NamedTuple):
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: > Similarly, the Python class `FlowLayout` clashes with a Java-side layout manager of the same name. How are other `QLayout` subclasses named? And was there really no built-in flow-layout-like thing in Qt already?

I guess they're called `Q<Something>Layout`. We can rename it to `QFlowLayout`, I suppose...

And nope, there isn't. What they do have is [an example for creating one](https://doc.qt.io/qt-6/qtwidgets-layouts-flowlayout-example.html), which is where I think I got our flow layout from...

File: src/napari_imagej/widgets/focuser.py [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+            if len(actions) > 1 and QApplication.keyboardModifiers() & Qt.ShiftModifier:
+                actions[1].action()
+            else:
+                actions[0].action()
+
+    def _python_actions_for(
+        self, result: "jc.SearchResult"
+    ) -> Dict[str, List[SearchAction]]:
+        """
+        Gets the list of predefined button parameters that should appear
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Well I just added docstrings in 3b8a65e.

File: src/napari_imagej/widgets/focuser.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Renamed `FlowLayout` to `QFlowLayout` in 4023e34

File: src/napari_imagej/widgets/searchbar.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/searchbar.py]: And I then proceeded to rename `SearchActionDisplay` to `ResultRunner` in 3774e20.

File: src/napari_imagej/widgets/focuser.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Removed `SearchAction` completely in 8841d6e6

File: src/napari_imagej/widgets/focuser.py (hunk 3 above)
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: It's not called anything close to this anymore!

File: tests/widgets/test_results_table.py (hunk 5 above)
Comment (gselzer) [reply tests/widgets/test_results_table.py]: Renamed to something completely different!

File: src/napari_imagej/widgets/focuser.py [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+
+    tooltips: Dict[str, str] = {
+        "Widget": "Runs functionality from a napari widget. "
+        "Useful for parameter sweeping",
+        "Run": "Runs functionality from a modal widget. Best for single executions",
+        "Source": "Opens the source code on GitHub",
+        "Help": "Opens the functionality's ImageJ.net wiki page",
+    }
+
+    def _actions_from_result(self, result: "jc.SearchResult") -> List[SearchAction]:
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I made some changes earlier that resulted in `_actions_from_result` being renamed to `buttons_for`, so I just changed ` convert_searchResult_to_info` to `info_for`.

File: src/napari_imagej/widgets/focuser.py [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+                        ij().py.from_java(result.name()),
+                        convert_searchResult_to_info(result),
+                        modal=True,
+                    ),
+                ),
+                SearchAction(
+                    name="Widget",
+                    action=lambda: self._execute_module(
+                        ij().py.from_java(result.name()),
+                        convert_searchResult_to_info(result),
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Renamed to `info_for` in 1e3c045

File: src/napari_imagej/widgets/focuser.py [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+                        modal=False,
+                    ),
+                ),
+            ],
+        }
+
+    tooltips: Dict[str, str] = {
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Sure. See scijava/scijava-search#26

File: src/napari_imagej/widgets/focuser.py [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+        "Widget": "Runs functionality from a napari widget. "
+        "Useful for parameter sweeping",
+        "Run": "Runs functionality from a modal widget. Best for single executions",
+        "Source": "Opens the source code on GitHub",
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Rewritten in 9321be4

File: src/napari_imagej/widgets/focuser.py [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+                        modal=False,
+                    ),
+                ),
+            ],
+        }
+
+    tooltips: Dict[str, str] = {
+        "Widget": "Runs functionality from a napari widget. "
+        "Useful for parameter sweeping",
+        "Run": "Runs functionality from a modal widget. Best for single executions",
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Rewritten in 9321be4

File: src/napari_imagej/widgets/focuser.py [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+                        ij().py.from_java(result.name()),
+                        convert_searchResult_to_info(result),
+                        modal=False,
+                    ),
+                ),
+            ],
+        }
+
+    tooltips: Dict[str, str] = {
+        "Widget": "Runs functionality from a napari widget. "
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Rewritten in 9321be4

File: src/napari_imagej/widgets/focuser.py [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+    def _actions_from_result(self, result: "jc.SearchResult") -> List[SearchAction]:
+        button_params: List[SearchAction] = []
+        # Get all additional python actions for result
+        python_action_replacements: Dict[str, SearchAction] = self._python_actions_for(
+            result
+        )
+        # Iterate over all available python actions
+        searchService = ij().get("org.scijava.search.SearchService")
+        for java_action in searchService.actions(result):
+            action_name = ij().py.from_java(java_action.toString())
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I fixed this...somewhere.

File: src/napari_imagej/widgets/focuser.py [hunk 34]
```diff
@@ -0,0 +1,174 @@
...
+            # If we have python replacements for this action, use them
+            if action_name in python_action_replacements:
+                button_params.extend(python_action_replacements[action_name])
+            # Otherwise, wrap the java action into a python action
+            else:
+                params = SearchAction(name=action_name, action=java_action.run)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: With the whole Widget simplified, we don't have this code anymore.

File: src/napari_imagej/widgets/focuser.py [hunk 35]
```diff
@@ -0,0 +1,174 @@
...
+                button_params.append(params)
+        return button_params
+
+    def clear_focus(self):
+        self.setText("")
+        # Hide buttons
+        for button in self.focused_action_buttons:
+            button.hide()
+
+    def focus(self, result: "jc.SearchResult"):
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I did it!

File: src/napari_imagej/widgets/focuser.py [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+        name = ij().py.from_java(result.name())  # type: ignore
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: Fixed it somewhere.

File: src/napari_imagej/widgets/focuser.py [hunk 37]
```diff
@@ -0,0 +1,174 @@
...
+        self.setText(name)
+
+        # Create buttons for each action
+        # searchService = ij().get("org.scijava.search.SearchService")
+        # self.focused_actions = searchService.actions(module)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: They're gone!

File: src/napari_imagej/widgets/focuser.py [hunk 38]
```diff
@@ -0,0 +1,174 @@
...
+            else:
+                params = SearchAction(name=action_name, action=java_action.run)
+                button_params.append(params)
+        return button_params
+
+    def clear_focus(self):
+        self.setText("")
+        # Hide buttons
+        for button in self.focused_action_buttons:
+            button.hide()
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I dunno. All I know is that this code no longer exists!

File: src/napari_imagej/widgets/focuser.py [hunk 39]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Create buttons for each action
+        # searchService = ij().get("org.scijava.search.SearchService")
+        # self.focused_actions = searchService.actions(module)
+        python_actions: List[SearchAction] = self._actions_from_result(result)
+        buttons_needed = len(python_actions)
+        activated_actions = len(self.focused_action_buttons)
+        # Hide buttons if we have more than needed
+        while activated_actions > buttons_needed:
+            activated_actions = activated_actions - 1
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: This code is gone!

File: src/napari_imagej/widgets/focuser.py [hunk 40]
```diff
@@ -0,0 +1,174 @@
...
+    def focus(self, result: "jc.SearchResult"):
+        name = ij().py.from_java(result.name())  # type: ignore
+        self.setText(name)
+
+        # Create buttons for each action
+        # searchService = ij().get("org.scijava.search.SearchService")
+        # self.focused_actions = searchService.actions(module)
+        python_actions: List[SearchAction] = self._actions_from_result(result)
+        buttons_needed = len(python_actions)
+        activated_actions = len(self.focused_action_buttons)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: This code is gone.

File: src/napari_imagej/widgets/focuser.py [hunk 41]
```diff
@@ -0,0 +1,174 @@
...
+        # Hide buttons if we have more than needed
+        while activated_actions > buttons_needed:
+            activated_actions = activated_actions - 1
+            self.focused_action_buttons[activated_actions].hide()
+        # Create buttons if we need more than we have
+        while len(self.focused_action_buttons) < buttons_needed:
+            button = QPushButton()
+            self.focused_action_buttons.append(button)
+            self.button_pane.layout().addWidget(button)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/focuser.py]: I don't know why I didn't do it before, but I'm not doing it anymore!

File: src/napari_imagej/widgets/results.py [hunk 42]
```diff
@@ -0,0 +1,206 @@
...
+from jpype import JArray, JImplements, JOverride
+from qtpy.QtCore import Qt, Signal, Slot
+from qtpy.QtWidgets import QTreeWidget, QTreeWidgetItem
+
+from napari_imagej.setup_imagej import ensure_jvm_started, ij, jc
+
+
+class ResultTreeItem(QTreeWidgetItem):
+    """
+    A QTreeWidgetItem wrapping a org.scijava.search.SearchResult
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Added in 112e8d5

File: src/napari_imagej/widgets/results.py [hunk 43]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self, result: "jc.SearchResult"):
+        super().__init__()
+        self.name = ij().py.from_java(result.name())
+        self.result = result
+
+        # Set QtPy properties
+        self.setText(0, self.name)
+
+
+class SearchEventWrapper:
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Or, even better: get rid of `SearchEventWrapper` entirely!

See cbc3746

File: src/napari_imagej/widgets/results.py [hunk 44]
```diff
@@ -0,0 +1,206 @@
...
+
+    # -- Helper Functionality -- #
+
+    def _init_producer(self):
+        # First, wait for the JVM to start up
+        ensure_jvm_started()
+
+        # Then, define our SearchListener
+        @JImplements("org.scijava.search.SearchListener")
+        class NapariSearchListener:
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Well, for consistentcy, `NapariImageJSearchListener`

[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Renamed in 25e2e36

File: src/napari_imagej/widgets/results.py [hunk 45]
```diff
@@ -0,0 +1,206 @@
...
+
+        # Start the search!
+        # NB: SearchService.search takes varargs, so we need an array
+        listener_arr = JArray(jc.SearchListener)([NapariSearchListener(self.process)])
+        self._searchOperation = (
+            ij().get("org.scijava.search.SearchService").search(listener_arr)
+        )
+        # Make sure that the search stops when we close napari
+        # Otherwise the Java threads like to continue
+        atexit.register(self._searchOperation.terminate)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Good question. Maybe we should instead use `when_jvm_stops`.

[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Fixed in 6b92451

File: src/napari_imagej/setup_imagej.py [hunk 46]
```diff
@@ -88,23 +124,7 @@ def running_headless() -> bool:
...
-    """
-    Returns the ImageJ instance.
-    If it isn't ready yet, blocks until it is ready.
-    """
-    return ij_instance.get()
-
-
-def logger():
-    return _logger
+ij_instance: AsyncResult = threadpool.apply_async(func=_imagej_init)
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/setup_imagej.py]: Renamed with 7319cfb9

File: src/napari_imagej/widgets/results.py (hunk 45 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: Fixed with 6b924516

File: src/napari_imagej/napari.yml [hunk 47]
```diff
@@ -3,7 +3,7 @@ display_name: napari-imagej
 contributions:
   commands:
     - id: napari-imagej.func
-      python_name: napari_imagej.widget:ImageJWidget
+      python_name: napari_imagej.widgets.napari_imagej:NapariImageJ
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/napari.yml:6]: Cleaned up in a9828ba

File: src/napari_imagej/__init__.py [hunk 48]
```diff
@@ -1,8 +1,28 @@
+"""
+napari-imagej provides a graphical wrapper for the PyImageJ project, exposed as a
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/__init__.py]: Cleaned up in https://github.com/imagej/napari-imagej/commit/a9828ba08507e19d286ea7e33f201c69b16a97f6

File: src/napari_imagej/napari.yml (hunk 47 above)
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/napari.yml]: Renamed to `NapariImageJWidget` in e4f9612

File: src/napari_imagej/types/placeholders.py (hunk 6 above)
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: We do this now! See [1212ec5](https://github.com/imagej/napari-imagej/pull/95/commits/1212ec5d03d4117ea85537bc14bb4f93c3f2624c)

File: src/napari_imagej/types/placeholders.py [hunk 49]
```diff
@@ -0,0 +1,76 @@
...
+"""
+from enum import Enum, auto
+from typing import List
+
+from napari_imagej.setup_imagej import jc
+
+
+class PythonStandin(Enum):
```
[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: Well, so now there are two things, which should be treated separately. The enums and the enum-likes.

We now have all enums covered automatically.

So now we just have to figure out how to best deal with the enum-likes...

[the remark this answers is not in the document]
Comment (gselzer) [reply src/napari_imagej/types/placeholders.py]: Renamed to `JavaEnumLike`s.

File: src/napari_imagej/types/converters/__init__.py (hunk 7 above)
Comment (gselzer) [reply src/napari_imagej/types/converters/__init__.py]: Removed in e3a5ce8

File: src/napari_imagej/widgets/results.py (hunk 8 above)
Comment (gselzer) [reply src/napari_imagej/widgets/results.py]: I think all of this has been renamed.

* `SearchResultTree` is the main tree
* `SearcherTreeItem` is the non-leaf node containing a `Searcher`
* `SearchResultTreeItem` is the leaf node class containing a `SearchResult`.

**Pull request merged**